packages/astro/src/runtime/server/index.ts

```diff
@@ -106,7 +106,8 @@ export async function renderComponent(result: SSRResult, displayName: string, Co
     return output;
   }
 
-  let metadata: AstroComponentMetadata = { displayName };
+  const metadata: AstroComponentMetadata = { displayName };
+  const assets: AstroComponentAssets = { styles: result.styles, scripts: result.scripts, useHydrationScript: false };
```

Comment:
ohhh, this is very nice. Since this is passing along `result.scripts` it should only add them once ever.

Reply:
Just noting that the downside to this approach is that we are giving renderers more control. We don't want them adding arbitrary scripts when the user has not opted in to client rendering.

Maybe a solution is another hook on the renderer that we only call on components that are going to be hydrated. Then the renderer can add scripts, but only in that scenario. So maybe not passing `assets` to `renderToStaticMarkup` but rather to this new hook.

This is why this would be good as its own separate RFC. Wouldn't want specifics of that change to block the renderer idea (or vice-versa).

Reply:
> We don't want them adding arbitrary scripts when the user has not opted in to client rendering.

In the case of custom elements, I think this paradigm is shifted. Custom Elements are intended to be served statically as-is. For instance, unlike any other library, Custom Elements can change the structure surrounding their contents during hydration without disconnecting any of the contents from the DOM. To that end, authors are expected to account for both the defined and non-defined (pre-hydrated) states.

There is a special CSS property just for this. https://developer.mozilla.org/en-US/docs/Web/CSS/:defined

Reply:
It's true that most custom elements are designed for client-side usage (so are most framework components, fwiw), but I don't think that changes are stance on JS being opt-in with Astro. We can't (easily) stop renderers from adding scripts to the page, but we also don't have to make it easier than it should be.

Reply:
Ah! This is actually why the return interface of renderers is `{ html }`. The idea was that should they ever need to bubble up scripts or styles, they could also return `{ css, js }`! That way we're not exposing `result` directly to the renderers, but internally that's where we would put them.

Reply:
> The idea was that should they ever need to bubble up scripts or styles, they could also return `{ css, js }`

That sounds wonderful.

1. Is that already an approved functionality or does it need an RFC?
2. Should that be implemented _before_ this PR or RFC could be accepted, or should this pass/fail as-is and the implementation change later?

Reply:
Will need an RFC.

Reply:
The reason for all of the `client:` directives is to make the user think about the best way for a component to load. That why none of them are the default. Allowing renderers to subvert that would take away from one of Astros biggest features, in my opinion.

Reply:
@matthewp, here it is. https://github.com/snowpackjs/astro/issues/1766

This was a very difficult RFC to write, because renders and client directives are built on a premise that components are 'single use only'. I now think the current `client:` directives implementation is flawed, as any component could be used on the same page with different client directives. While I might expect the earliest matching directive to 'win', in reality each component probably handles them independently. This flaw is exemplified by Lit components, since the registration of one custom element immediately affects all other elements on the page.